image_proc/src/crop_decimate.cpp

```diff
@@ -0,0 +1,315 @@
+      offset_x_ == 0                  &&
+      offset_y_ == 0                  &&
+      width  == (int)image_msg->width &&
+      height == (int)image_msg->height)
+  {
+    pub_.publish(image_msg, info_msg);
+    return;
+  }
+
+  // Get a cv::Mat view of the source data
+  CvImageConstPtr source = toCvShare(image_msg);
+
+  // Except in Bayer downsampling case, output has same encoding as the input
+  CvImage output(source->header, source->encoding);
+  // Apply ROI (no copy, still a view of the image_msg data)
+  output.image = source->image(cv::Rect(offset_x_, offset_y_, width, height));
+
+  // Special case: when decimating Bayer images, we first do a 2x2 decimation to BGR
+  if (is_bayer && (decimation_x > 1 || decimation_y > 1))
+  {
+    if (decimation_x % 2 != 0 || decimation_y % 2 != 0)
+    {
+      RCLCPP_ERROR(get_logger(),
+        "Odd decimation not supported for Bayer images");
+      return;
+    }
+
+    cv::Mat bgr;
+    int step = output.image.step1();
+    if (image_msg->encoding == sensor_msgs::image_encodings::BAYER_RGGB8)
+      debayer2x2toBGR<uint8_t>(output.image, bgr, 0, 1, step, step + 1);
+    else if (image_msg->encoding == sensor_msgs::image_encodings::BAYER_BGGR8)
+      debayer2x2toBGR<uint8_t>(output.image, bgr, step + 1, 1, step, 0);
+    else if (image_msg->encoding == sensor_msgs::image_encodings::BAYER_GBRG8)
+      debayer2x2toBGR<uint8_t>(output.image, bgr, step, 0, step + 1, 1);
+    else if (image_msg->encoding == sensor_msgs::image_encodings::BAYER_GRBG8)
+      debayer2x2toBGR<uint8_t>(output.image, bgr, 1, 0, step + 1, step);
+    else if (image_msg->encoding == sensor_msgs::image_encodings::BAYER_RGGB16)
+      debayer2x2toBGR<uint16_t>(output.image, bgr, 0, 1, step, step + 1);
+    else if (image_msg->encoding == sensor_msgs::image_encodings::BAYER_BGGR16)
+      debayer2x2toBGR<uint16_t>(output.image, bgr, step + 1, 1, step, 0);
+    else if (image_msg->encoding == sensor_msgs::image_encodings::BAYER_GBRG16)
+      debayer2x2toBGR<uint16_t>(output.image, bgr, step, 0, step + 1, 1);
+    else if (image_msg->encoding == sensor_msgs::image_encodings::BAYER_GRBG16)
+      debayer2x2toBGR<uint16_t>(output.image, bgr, 1, 0, step + 1, step);
+    else
+    {
+      RCLCPP_ERROR(get_logger(), "Unrecognized Bayer encoding '%s'",
+        image_msg->encoding.c_str());
+      return;
+    }
+
+    output.image = bgr;
+    output.encoding = (bgr.depth() == CV_8U) ? sensor_msgs::image_encodings::BGR8
+                                             : sensor_msgs::image_encodings::BGR16;
+    decimation_x /= 2;
+    decimation_y /= 2;
+  }
+
+  // Apply further downsampling, if necessary
+  if (decimation_x > 1 || decimation_y > 1)
+  {
+    cv::Mat decimated;
+
+    if (interpolation_ == image_proc::CropDecimate_NN)
+    {
+      // Use optimized method instead of OpenCV's more general NN resize
+      int pixel_size = output.image.elemSize();
+      switch (pixel_size)
```

Comment:
Why not just call `decimate<pixel_size>` and let `decimate` log the error if it's an unsupported size? It'll save a lot of code.

Reply:
This was just directly ported from ROS1, but from looking at it, not all the options are ordered numeric, so its a way of enforcing the use of specific values. I agree it could be done other ways

Reply:
I'm OK with leaving it for later, though.